runtime/src/main/java/org/corfudb/runtime/view/SequencerView.java

```diff
@@ -31,17 +34,47 @@ public SequencerView(CorfuRuntime runtime) {
      * @return The first token retrieved.
      */
     public TokenResponse nextToken(Set<UUID> streamIDs, int numTokens) {
-        return layoutHelper(e -> CFUtils.getUninterruptibly(e.getPrimarySequencerClient()
-                .nextToken(streamIDs, numTokens)));
+        if (numTokens == 0) {
+            long tail = checkHints(streamIDs);
+            if (tail != Address.NOT_FOUND) {
+                // the request can be serviced from sequencer hints (omitting sequencer request)
+                long epoch = TransactionalContext.getCurrentContext().hintsEpoch;
+                return new TokenResponse(tail, epoch, null);
```

Comment:
It seems that returning a TokenResponse with some special epoch indicating that it was retrieved from a hint would be a better idea. Either way it is not used or verified by downstream code in the stream layer (the only place the epoch is checked is AddressSpaceView::write, and copy, but that is irrelevant).